src/endless/Eosldr.cpp

```diff
@@ -0,0 +1,287 @@
+            IFFALSE_PRINTERROR(SetFileAttributes(m_path, m_originalAttributes),
+                "Couldn't restore attributes");
+        }
+    }
+
+    bool succeeded() { return m_succeeded; }
+    DWORD originalAttributes() { return m_originalAttributes; }
+
+private:
+    const CString &m_path;
+    DWORD m_originalAttributes;
+    bool m_succeeded;
+};
+
+bool EosldrInstallerBootIni::ModifyBootOrder(const CString & systemDriveLetter, const CString & eosldrMbrPath, bool add, bool &foundBootEntry)
+{
+    const CString bootIni = systemDriveLetter + BOOT_INI;
+    FUNCTION_ENTER_FMT("bootIni=%ls", bootIni);
+
+    Unveil unveilBootIni(bootIni);
+    IFFALSE_RETURN_VALUE(unveilBootIni.succeeded(), "Couldn't make boot.ini visible", false);
+
+    const CString bootIniBackup = systemDriveLetter + _T(BOOT_INI_BACKUP);
+    IFFALSE_PRINTERROR(INVALID_FILE_ATTRIBUTES != SetFileAttributes(bootIniBackup, FILE_ATTRIBUTE_NORMAL),
+        "Couldn't make " BOOT_INI_BACKUP " visible (it probably doesn't exist)");
+    if (!CopyFile(bootIni, bootIniBackup, /* bFailIfExists */ FALSE)) {
+        PRINT_ERROR_MSG_FMT("Backing up boot.ini with CopyFile(%ls, %ls, FALSE) failed",
+            bootIni, bootIniBackup);
+        return false;
+    }
+
+    IFFALSE_PRINTERROR(SetFileAttributes(bootIniBackup, unveilBootIni.originalAttributes()),
+        "SetFileAttributes failed on backup");
+
+    if (add) {
+        const CString quotedName = L'"' + m_name + L'"';
+        if (!WritePrivateProfileString(BOOT_INI_SECTION, eosldrMbrPath, quotedName, bootIni)) {
+            PRINT_ERROR_MSG_FMT("WritePrivateProfileString(..., %ls, %ls, %ls) failed",
+                eosldrMbrPath, quotedName, bootIni);
+            return false;
+        }
+    } else {
+        // If there's an existing entry for eosldr with a non-empty-string label,
+        // consider the boot entry to be found.
+        wchar_t dummy[2] = { 0 };
+        DWORD len = GetPrivateProfileString(BOOT_INI_SECTION, eosldrMbrPath, L"", dummy, ARRAYSIZE(dummy), bootIni);
+        foundBootEntry = (len > 0);
+
+        // passing NULL for the third parameter removes the entry
+        if (!WritePrivateProfileString(BOOT_INI_SECTION, eosldrMbrPath, NULL, bootIni)) {
+            PRINT_ERROR_MSG_FMT("WritePrivateProfileString(..., %ls, NULL, %ls) failed",
+                eosldrMbrPath, bootIni);
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool EosldrInstallerBcd::AddToBootOrder(const CString & systemDriveLetter, const CString & eosldrMbrPath)
+{
+    FUNCTION_ENTER_FMT("%ls, %ls", systemDriveLetter, eosldrMbrPath);
+
+    CString guid;
+    IFFALSE_RETURN_VALUE(
+        WMI::AddBcdEntry(m_name, eosldrMbrPath, guid),
+        "AddBcdEntry failed", false);
+
+    // At this point it's too late to fail: we've already added the entry to the boot order.
+    // TODO: pass GUID into ::AddBcdEntry instead, and add it to the registry *first*
```

Comment:
I have tested a VM with two hard disks:

* HD0: boot device, Windows 7
* HD1: Windows 8

and added the Windows 8 installation on HD1 to the BCD of HD0.

Assuming neither HD has Endless installed, then if I boot either copy of Windows and install Endless OS, it gets installed to the booted Windows' HD, added to the correct BCD (on HD0), and loads correctly from the menu.

If I install it to both at once, I see two entries in the boot menu (correct) but the copy on HD0 is always booted. I assume this is because `$bootdev` is always `hd0` so `search.file` always finds the installation on that disk. It might be possible to arrange for `eosldr.mbr` to update whatever `$bootdev` is based on, at the cost of obscuring which BCD is in use(?).

The other option is to only support one installation at a time, in which case we can indeed bake in a GUID here.